pkg/core/blockchain.go

```diff
@@ -682,29 +682,57 @@ func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage)
 	v := bc.dao.Version
 	cache := bc.dao // dao is MemCachedStore over DB, so use dao directly to persist cached changes right to the underlying DB
 
-	bc.log.Info("initialize state reset", zap.Uint32("target height", height))
+	bc.log.Info("initializing state reset", zap.Uint32("target height", height))
 	start := time.Now()
 	p := start
-	keys := 0
+
+	// Start batch persisting routine, it will be used for blocks/txs/AERs/storage items batches persist.
+	type postPersist func(persistedKeys int, err error) error
+	var (
+		persistCh       = make(chan postPersist)
+		persistToExitCh = make(chan struct{})
+	)
+	go func() {
+		for {
+			f, ok := <-persistCh
+			if !ok {
+				break
+			}
+			persistErr := f(cache.Persist())
```

Comment:
One thing I'm not excited about in this approach is that this line can be executed at any time, so we may have one stage finished and then some additional changes from the next stage (it works with the same `cache`). If the node is to crash there is a potential for inconsistent state.

This can be avoided if we're to use new DAO for every stage and then synchronize before flushing the new portion of data into the `cache.`